### PR TITLE
heavy COBOL update, first support within function list and auto-completion

### PR DESCRIPTION
**Pull request by GitMensch**
COBOL support in function list - new entries in functionList.xml (note: defaults to fixed-form reference format, manual switch to free-form reference-format possible in association for langID 50)

COBOL support in auto-completion - new file APIs/cobol.xml (note: defaults to upper-case, hopefully changed to support both, see https://github.com/notepad-plus-plus/notepad-plus-plus/issues/3998)
COBOL syntax highlighter - recreated list with many more reserved words

note: list of statements, functions and reserved words token from GnuCOBOL 3.0 (COBOL85,2002,2014 + extensions from IBM/MicroFocus/RM/ACUCOBOL), please report back if any words are missing

### CODE REVIEW

**Comment by GitMensch** [comment]
Note: While the auto-complete works for the *keyword* [`FORMATTED-DATE`](https://github.com/GitMensch/notepad-plus-plus/blob/d53f6c596dfc90e110ae9ef9957807eaf2800e09/PowerEditor/installer/APIs/cobol.xml#L415) it doesn't work for the *function* parameters. Is there anything I can do to fix this or should I file a bug?

**Comment by chcg** [comment]
@MAPJe71 Maybe interesting for  #3393

**Comment by dail8859** [comment]
@GitMensch 

I have experienced the opposite when using `additionalWordChar`, for example using `.` for Lua. Using `Ctrl+Space` does not suggest words with the character already in them (see #2532), however opening the function (e.g. with `(`) works correctly. See the gif below. After it auto-completes the full function I delete a few chars and try to trigger the auto complete again but it doesn't work, however if I open the function it works correctly. I did not try any changes in this PR.


acdemo

File: PowerEditor/src/functionList.xml [hunk 1]
```diff
@@ -90,6 +90,11 @@
 			<association id=  "powershell_function"   langID="53"                          />
 			<association id=  "javascript_function"   langID="58"                          />
 			<association id=       "baanc_section"    langID="60"                          />
+			<!--
+			   As there is currently only one langID for COBOL: change to cobol_section_free
+			   if this is your favourite format
+			-->
+			<association langID="50" id="cobol_section_fixed"/>
```
Comment (MAPJe71) [inline PowerEditor/src/functionList.xml]: Use same layout as is used for already available associations and position based on `langID` value i.e. between "innosetup_syntax" and "powershell_function" associations.

File: PowerEditor/src/functionList.xml [hunk 2]
```diff
@@ -1468,6 +1473,28 @@
...
+			<!-- Variant for COBOL fixed-form reference format -->
+			<parser id="cobol_section_fixed" displayName="COBOL fixed-form reference format">
+				<function
+					mainExpr="(?m-s)^.{6}[ D][\t ]{0,3}(?!exit\s)[\w_-]+(\.|((?'seps'([\t ]|\*&gt;.*|([\n\r]+(.{6}([ D]|\*.*)|.{0,6}$)))+)section(\.|((?&amp;seps)(\.|[\w_-]+\.)))))"
+					displayMode="$functionName">
+					<functionName>
+						<nameExpr expr="[\w_-]+((?=\.)|((?'seps'([\t ]|\*&gt;.*|([\n\r]+(.{6}([ D]|\*.*)|.{0,6}$)))+)section((?=\.)|(?&amp;seps)((?=\.)|[\w_-]+(?=\.)))))"/>
+					</functionName>
+				</function>
+			</parser>
```
Comment (MAPJe71) [inline PowerEditor/src/functionList.xml]: 1. Use same layout as is used for existing parsers.
2. Remove `displayMode` attribute, it's not used.
3. Add comment/description.

File: PowerEditor/src/functionList.xml [hunk 3]
```diff
@@ -1468,6 +1473,28 @@
...
+			<!-- Variant for COBOL free-form reference format -->
+			<parser id="cobol_section_free" displayName="COBOL free-form reference format">
+				<function
+					mainExpr="[\s\.](?!exit\s)[\w_-]+\s+section(\s*|(\s+[\w_-]+)?)(?=\.)"
+					displayMode="$functionName">
+					<functionName>
+						<nameExpr expr="[\w_-]+\s*section"/>
+				   </functionName>
+				</function>
+			</parser>
```
Comment (MAPJe71) [inline PowerEditor/src/functionList.xml]: See review comment for fixed-form parser.

**Comment by GitMensch** [comment]
@MAPJe71 Applied most of your comments. Note: We tweaked these entries back in 2015 together, I would actually need time to fully understand and explain these regexes. But as the functionList still doesn't support COBOL at all I consider it better to use these as they are than waiting for me to come back to them in 2018/2019. Note: I don't know if the old comment boundaries problem is fixed by now (there was a request for you to redo your patch back then), if they are the commented one for free-form reference-format may be retested.

... just rechecked:
https://svn.code.sf.net/p/open-cobol/contrib/trunk/samples/prime_machine/prime_machine.cob works fine with the "simple" cobol_section_free variant which is active in the suggested PR but not with the extended one including the paragraphs.

**Comment by GitMensch** [comment]
@donho The "common style issues" mentioned by @MAPJe71 were fixed, all CI and manual checks passed and #2731 is fixed by this - is there anything "open" on your side before including this one?

**Pull request closed without merging**